### DISCUSSION TITLE
Exposing Tailwind marketing and misinformation engine

### DISCUSSION THREAD

**Discussion by tipiirai** [nuejs/nue#207] [Announcements]
Please check out perhaps the most important content piece I've ever written:

[Tailwind marketing and misinformation engine](https://nuejs.org/blog/tailwind-misinformation-engine/)

<img src="https://nuejs.org/img/tailwind-messaging-pillars.jpg">

**Comment by nobkd** [comment]
I think you have a typo there:

https://nuejs.org/blog/tailwind-misinformation-engine/#phase2

> Once they **stared** cashing, Tailwind wanted to make sure the users were properly onboarded and locked in to the system.

started

---

Ok, found more

https://nuejs.org/blog/tailwind-misinformation-engine/#i-love-css

> The **nyanced** relationship between [form and function](https://en.wikipedia.org/wiki/Form_follows_function).

nuanced

---

https://nuejs.org/blog/tailwind-misinformation-engine/#first-things-first-learn-css

> Study MDN documentation on web standards. There's a lot, so start with the most important aspects of CSS: [the cascade](https://developer.mozilla.org/en-US/docs/Web/CSS/Cascade) and **[specifity](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_nesting/Nesting_and_specificity)**.

specificity

I also always get that wrong... ;)

---

https://nuejs.org/blog/tailwind-misinformation-engine/#stay-relevant

> The utility soup produced today will eventually turn into a technical **dept**.

debt

---

Top of the page: https://nuejs.org/blog/tailwind-misinformation-engine/

> Tailwind CSS was born out of this sentence:

> "The most reusable components are those with class names that are independent of the content"

Maybe add the text in the picture to the alt text of the image?

**Reply by tipiirai** [reply]
Thank you @nobkd ! I fixed all the above issues

**Reply by nobkd** [reply]
> It was a gold mine for **Talwind's** commercial business model.

Tailwind's

Right **above** https://nuejs.org/blog/tailwind-misinformation-engine/#phase2

---

https://nuejs.org/blog/tailwind-misinformation-engine/#first-things-first-learn-css

> Start from the **Nicholas's** post and learn the benefits of semantic naming.

Nicholas'

**Reply by nobkd** [reply]
Oh. Will you update the [www repo](https://github.com/nuejs/www/blob/master/blog/tailwind-misinformation-engine/index.md) to include the up-to-date version of the post?

**Reply by tipiirai** [reply]
Done. Thanks!

**Comment by felix-bohlin** [comment]
First of all, thanks for a very interesting read. I have some thoughts though.

I haven't gotten the impression that the Tailwind team's intentions are ill-willed and cash grabby. My take on it is that  there's this camp of people that aren't as proficient in CSS and/or think writing CSS is the worst, and TW is primarily catered towards them or at least has gotten a lot of traction with those devs. They've probably been shielded from writing CSS before, but now that every company seemingly prefers to hire fullstack over BE/FE devs they can't get away anymore. 

For them, CSS has just been this weird _thing_ full of clearfixes, floats and seemingly contradictory properties that "don't add any new features apart from changing background color". On top of that it's another mental model! Then TW comes along, and I totally get it - as you said - writing it with HMR feels fast and satisfying, it holds your hand along the way with its IDE plugin and gorgeous documentation. When has CSS _ever_ done that for you (rethorical question)? 

Catalyst UI - yeah it's not beautiful but if you're writing React and don't pick MUI as your component library you have a harder time recruiting. CUI with the weight of TW behind it is going head-to-head with MUI, leaning even heavier into the arms of those fullstack devs, getting them to style components without writing much CSS.

For all the downsides of TW, I don't view see it as a machine to lock people in to squeeze them for money. I'm not saying this approach is good or bad, but I feel they're trying to solve a **systemic problem** more than anything else. Devs are expected to do everything now in half the time. Along comes TW, you're given a hammer and you're being told every problem is a nail - who wouldn't love that?

It's not all bad though and as people have mentioned before in this repo - Anthony Fu's UnoCSS has come along as a great middleground with a little bit of everything baked in. In my opinion, TW is most of all a beginning of a broader conversation, and who knows - maybe your upcoming CSS solution will have a say too? 🙂

**Reply by tipiirai** [reply]
Thank you for the great feedback! I can definitely see how Tailwind benefits people and they can move fast by just copy/pasting components and never worry what's under the hood. I just don't like the way Adam talks about CSS with terms like "fucking bullshit" and the origin story out of that Nicholas' article is not cool. Seems I'm not alone here: just read [this article](https://bastianallgeier.com/notes/tailwind) from the creator of Derby. Or what do you think about that stuff?

**Reply by nobkd** [reply]
@tipiirai the link is to nuejs.org

**Reply by tipiirai** [reply]
Updated the link. Thanks!

**Reply by felix-bohlin** [reply]
I see myself as being very active in keeping up with the web dev sphere but I might've missed the worst parts since I'm not on Twitter. That's why I was surprised to read all that and I wonder if I've missed something completely, because I haven't noticed any bullish behaviour in their marketing, at all. When he says "fucking bullshit" he talks about the mess his team has created, not CSS as a whole (at least the way I hear it). The frustrations in regards to the complete lack of guardrails in CSS is understandable and his examples are great, because we've all been there honestly 😅 `class="card card--medium card--blue"`, yep - been there done that! If something in their marketing has been out of line then obviously it's very sad that it has come to that - I just haven't picked up on anything.

Above anything though - and the reason I attempt to write long-form coherent comments here - is that I feel the underlying issue to all this often is overlooked. The reason I don't percieve him shitting on CSS is because I believe the issue Adam had wasn't tied to CSS, but to _communication_. We've all been there - how can I communicate to my team of variable CSS knowledge how to write maintainable CSS? As Bastian Allgeier mentions in the article you linked TW _excels_ in communication. It's the CSS with guardrails Adam always wanted that could keep his team from making their codebase "fucking bullshit" ever again. 

Setting aside upsides and downsides, Adam found a way to communicate that resonated with a lot of people, and I think that fact is very valuable in and of itself. He highlights that the non-opinionated nature of CSS has its downsides in terms of communication and we need to think about that. CSS custom properties, `@layer` and nesting are great examples of CSS addressing that.

I don't know, maybe I strayed off-topic here, apologise 😅

**Reply by tipiirai** [reply]
The problem is the "Best practices don't actually work" gospel which should have been "The best practices did not work for me" — just like the shitting part. It's the reason you are indirectly promoting their inline styling with phrases like " they're trying to solve a systemic problem". It's the reason we're having this conversation in the first place because he managed to change the prevailing practices for so many people. 

It's the reason I'm having a hard time talking about the standard MDN practices in vanilla CSS, like cascading, nesting, and all the cool new stuff that is on the specification. 

You see, the best practices have always worked for me. And I've never been on the `card--blue` side of things.

**Reply by felix-bohlin** [reply]
Thanks for clarifying that, I hear you!

The interesting question then is: what's the timeline here? When did we go wrong? What problems did we try to solve, or what chain of events lead to this gradual shift away from best practices?

Thanks for taking your time and chat a bit. Appreciate it.

**Reply by tipiirai** [reply]
Nice! 

I don't have all the answers, but I see that the turning point was when Facebook announced how they solved [all their issues](https://speakerdeck.com/vjeux/react-css-in-js) on their design system by switching from a PHP to a more advanced JavaScript setup. Namely global namespace pollution and dead code elimination. 

Developers accepted the new norm: "global namespace is the root of all evil", and rushed implementing different kinds of CSS-in-JS solutions. 

At the same time we forgot the fundamental principle of contemporary design: *Form follows Function*. You should build components based on their purpose, and not how they look. This principle has many forms including separation of concerns, or decoupled design, or the famous one from Steve Jobs "It's not how it looks, it's how it works".

For example: you have a table component, whose purpose is to show tabular data. On your front page, it may take a promotional form, but on the documentation area the purpose is to teach and educate — thus clarity is more important than the WOW effect. The same goes for other components, like tabbed interfaces or feedback forms. You get the idea. 

Now, when styling is jammed to the structure, you have a more or less hardcoded look & feel. And you need to create a new component to adapt the purpose. One good example is Catalyst and its many forms of the paragraph tag: Text, Description, DialogDescription, and AlertDescription. In an attempt to increase reusability, they reduced it. 

Summary: By solving a theoretical global namespace issue, we ended up breaking a fundamental design practice: Form follows function.

**Reply by scythewyvern** [reply]
what do you think a bad design can lead bad css?

i mean sometimes with a good design you can organise you styles with pleasure but sometimes bad design don't let you any chance.

so, should design go with development or development should figure out how to implement things in a better way?

**Reply by tipiirai** [reply]
I think the relationship between design and development is crucial and they both should go hand in hand. I think an easy-to-use design system helps both parties tremendously here, which is my next development focus. I'm planning to release a solid alternative to CSS-in-JS and Tailwind that helps you keep things in order. Hopefully out next month.

**Comment by felix-bohlin** [comment]
How are your blog analytics @tipiirai? Let's see if your article [gets mentioned](https://www.youtube.com/watch?v=_ErqxKGBYQQ) in a couple of hours 😉

**Reply by tipiirai** [reply]
👍 solid

**Reply by tipiirai** [reply]
I can give you the stats tomorrow

**Reply by felix-bohlin** [reply]
Yup, you're featured at 4:22:38.

**Reply by tipiirai** [reply]
Let me guess before I click on it. Everyone in the discussion, including the host is a Tailwind user/convert — so the discussion is about this "crazy argument" that I made.. Hope I'm wrong.. clicking now

**Reply by tipiirai** [reply]
Okay. 30 seconds was enough to see that the guy had picked up his side before reading the article. Even the beginning of the article, which explains how Tailwind got started _by the words of Adam_ was a wrong take. Not hard to imagine how the rest of the critique goes.

**Reply by joshamaju** [reply]
One good rule of thumb "never listen to Theo". And it's weird he thought the article was about him

**Reply by tipiirai** [reply]
Indeed. Just watched [this video](https://www.youtube.com/watch?v=s4BFIDYYYCA&t=2s) about him. Pretty alarming.

>  it's weird he thought the article was about him

Oh, wow. Haven't watched the video, but I'm not surprised.
